src/getDataFromTree.ts

```diff
@@ -129,51 +133,55 @@ export function walkTree(
 
       if (child) {
         if (Array.isArray(child)) {
-          child.forEach(item => walkTree(item, childContext, visitor));
+          child.forEach(item => walkTree(item, childContext, visitor, newContext));
         } else {
-          walkTree(child, childContext, visitor);
+          walkTree(child, childContext, visitor, newContext);
         }
       }
     } else if ((element.type as any)._context || (element.type as any).Consumer) {
       // A React context provider or consumer
-      if (visitor(element, null, context) === false) {
+      if (visitor(element, null, newContext, context) === false) {
         return;
       }
 
       let child;
-      if ((element.type as any)._context) {
+      if (!!(element.type as any)._context) {
         // A provider - sets the context value before rendering children
-        ((element.type as any)._context as any)._currentValue = element.props.value;
+        // this needs to clone the map because this value should only apply to children of the provider
+        newContext = new Map(newContext.entries());
+        newContext.set(element.type, element.props.value);
         child = element.props.children;
       } else {
         // A consumer
-        child = element.props.children((element.type as any)._currentValue);
+        child = element.props.children(
+          newContext.get((element.type as any).Provider) || (element.type as any)._currentValue,

```

Comment:
I think you need to be more specific with the check here:

```ts
let currentValue = (element.type as any)._currentValue;
if (childContext.subContexts.has((element.type as any).Provider)) {
  currentValue = childContext.subContexts.get((element.type as any).Provider);
}
child = element.props.children(currentValue);
```

In theory, someone could set the value of context to a falsy value, and that would be the value that should be respected. From the React docs:

> The defaultValue argument is only used by a Consumer when it does not have a matching Provider above it in the tree. This can be helpful for testing components in isolation without wrapping them. Note: passing undefined as a Provider value does not cause Consumers to use defaultValue.